modules/grandpa/src/benchmarking.rs

```diff
@@ -159,40 +125,14 @@ benchmarks_instance_pallet! {
 		assert!(<ImportedHeaders<T, I>>::contains_key(expected_hash));
 	}
 
-	// What we want to check here is the effect of many pre-commits on justification verification.
-	// We do this by creating many forks, whose head will be used as a signed pre-commit in the
-	// final justification.
-	submit_finality_proof_on_many_forks {
+	// This benchmark computes weight that needs to be added for every precommit. It assumes that all
+	// authorities have signed on the same header, so the cost of additional validator is near to
+	// the cost of validator signature verification.
+	weight_per_additional_precommit {
 		let p in 1..MAX_VALIDATOR_SET_SIZE;
-
+		let v = 1;
```

Comment:
Because there's another benchmark which is responsible for that. This benchmark must build `weight(p)`. And while building `weight(p)` I don't care about number of headers in the votes ancestry. So here I'm interested in additional weight that comes with additional signature verification + some of the loop checks. With that new custom justification validation there's a guarantee that every header in the ancestry will be only visited once (this doesn't include one additional visit per precommit, weight for which is added here). So we may use formula like that `a*A + b*B`. Benchmarks are responsible for computing `A`/`B`.

Reply:
So wasn't part of the initial problem here that different values of `p`/`v` would have different impacts depending on the size of the other variable?

E.g with `v=1` every additional `p` increases the weights by `X`, while with `v=1000` every additional `p` increases the weights by `X + C`?

Reply:
Yes. Take a look at the new verification code, please - all that happens with ancestry is: (1) every header is hashed and inserted into btree (this is accounted y the `weight_per_additional_unique_ancestor`) (2) zero or one header is accessed for every precommit (accounted by `weight_per_additional_precommit`) (3) every header in the valid justification is accessed at most one time (accounted by `weight_per_additional_unique_ancestor`). So while both parameters are still used inside one loop, it doesn't mean anymore that the weight formula should look something like `p*v*W`. Look at the attached doc - if what you're saying was the true, then the formula wouldn't fit that good for both small and large values of `(p, v)`.

Reply:
I agree with you, the formula shouldn't look like `p*v*W`. What it should look like is `A*p + B*v`. 

What I'm trying to say here is that the values of `A` and `B` will be different depending on the values of `p` and `v` used when running the benchmarks. Depending on if we fix `v=1` or `v=1000` the corresponding constants will be different - so what should we choose?